x/twap/strategy_test.go

```diff
@@ -0,0 +1,395 @@
+				osmoassert.DecApproxEq(suite.T(), test.expTwap, actualTwap, osmomath.GetPowPrecision())
+			}
+		})
+	}
+}
+
+// TestComputeArithmeticStrategyTwap tests computeArithmeticTwap on various inputs.
+// Contrary to computeTwap that handles the cases with zero delta correctly,
+// this function should panic in case of zero delta.
+func (suite *TwapStrategyTestSuite) TestComputeArithmeticStrategyTwap() {
+	pointOneAccum := OneSec.QuoInt64(10)
+	tests := map[string]computeTwapTestCase{
+		"basic: spot price = 1 for one second, 0 init accumulator": {
+			startRecord: newOneSidedRecord(baseTime, sdk.ZeroDec(), true),
+			endRecord:   newOneSidedRecord(tPlusOne, OneSec, true),
+			quoteAsset:  denom0,
+			expTwap:     sdk.OneDec(),
+		},
+		// this test just shows what happens in case the records are reversed.
+		// It should return the correct result, even though this is incorrect internal API usage
+		"invalid call: reversed records of above": {
+			startRecord: newOneSidedRecord(tPlusOne, OneSec, true),
+			endRecord:   newOneSidedRecord(baseTime, sdk.ZeroDec(), true),
+			quoteAsset:  denom0,
+			expTwap:     sdk.OneDec(),
+		},
+		"same record (zero time delta), division by 0 - panic": {
+			startRecord: newOneSidedRecord(baseTime, sdk.ZeroDec(), true),
+			endRecord:   newOneSidedRecord(baseTime, sdk.ZeroDec(), true),
+			quoteAsset:  denom0,
+			expPanic:    true,
+		},
+		"accumulator = 10*OneSec, t=5s. 0 base accum": testCaseFromDeltas(
+			suite, sdk.ZeroDec(), tenSecAccum, 5*time.Second, sdk.NewDec(2)),
+		"accumulator = 10*OneSec, t=3s. 0 base accum": testCaseFromDeltas(
+			suite, sdk.ZeroDec(), tenSecAccum, 3*time.Second, ThreePlusOneThird),
+		"accumulator = 10*OneSec, t=100s. 0 base accum": testCaseFromDeltas(
+			suite, sdk.ZeroDec(), tenSecAccum, 100*time.Second, sdk.NewDecWithPrec(1, 1)),
+
+		// test that base accum has no impact
+		"accumulator = 10*OneSec, t=5s. 10 base accum": testCaseFromDeltas(
+			suite, sdk.NewDec(10), tenSecAccum, 5*time.Second, sdk.NewDec(2)),
+		"accumulator = 10*OneSec, t=3s. 10*second base accum": testCaseFromDeltas(
+			suite, tenSecAccum, tenSecAccum, 3*time.Second, ThreePlusOneThird),
+		"accumulator = 10*OneSec, t=100s. .1*second base accum": testCaseFromDeltas(
+			suite, pointOneAccum, tenSecAccum, 100*time.Second, sdk.NewDecWithPrec(1, 1)),
+
+		"accumulator = 10*OneSec, t=100s. 0 base accum (asset 1)": testCaseFromDeltasAsset1(suite, sdk.ZeroDec(), OneSec.MulInt64(10), 100*time.Second, sdk.NewDecWithPrec(1, 1)),
+	}
+	for name, test := range tests {
+		suite.Run(name, func() {
+			osmoassert.ConditionalPanic(suite.T(), test.expPanic, func() {
+				arithmeticStrategy := &twap.ArithmeticTwapStrategy{ArithKeeper: *suite.App.TwapKeeper}
+				actualTwap := arithmeticStrategy.ComputeTwap(test.startRecord, test.endRecord, test.quoteAsset)
+				suite.Require().Equal(test.expTwap, actualTwap)
+			})
+		})
+	}
+}
+
+func (suite *TwapStrategyTestSuite) TestComputeGeometricStrategyTwap() {
```

Comment:
and this one.

These are the only test cases that test each implementation directly. The other test cases are general, and can be kept in their old files with minimal changes